### PR TITLE
Nexus worker support

### PR DESCRIPTION
**Pull request by Quinn-With-Two-Ns**
Nexus worker support to the Java SDK. This PR adds support for registering Nexus Services to Java SDK worker.

Note: This does **NOT** add support for _Nexus Operations -> Workflows_

### CODE REVIEW

File: temporal-sdk/src/main/java/io/temporal/worker/Worker.java [hunk 1]
```diff
@@ -386,11 +400,28 @@ public void registerActivitiesImplementations(Object... activityImplementations)
     workflowWorker.registerLocalActivityImplementations(activityImplementations);
   }
 
+  /**
+   * Register Nexus service implementation objects with a worker.
+   *
+   * <p>An Nexus service object must implement at least one interface annotated with {@link
+   * io.nexusrpc.handler.ServiceImpl}.
```
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/worker/Worker.java:407]:
```suggestion
   * <p>An Nexus service object must be annotated with {@link
   * io.nexusrpc.handler.ServiceImpl}.
```

Technically unrelated to interfaces

File: temporal-sdk/src/main/java/io/temporal/worker/Worker.java [hunk 2]
```diff
@@ -386,11 +400,28 @@ public void registerActivitiesImplementations(Object... activityImplementations)
...
+        "registerNexusServiceImplementation is not allowed after worker has started");
+    nexusWorker.registerNexusServiceImplementation(nexusServiceImplementations);
+  }
+
   void start() {
     if (!started.compareAndSet(false, true)) {
       return;
     }
     workflowWorker.start();
+    nexusWorker.start();
```
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/worker/Worker.java:424]: To save me from digging too much, can you confirm no resources are used by a Nexus worker if no Nexus service impls are registered?

File: temporal-sdk/src/main/java/io/temporal/worker/WorkflowImplementationOptions.java [hunk 3]
```diff
@@ -126,13 +129,39 @@ public Builder setDefaultLocalActivityOptions(
...
     }
 
+    /**
+     * Set individual Nexus Service options per service. Will be merged with the map from {@link
+     * io.temporal.workflow.Workflow#newNexusServiceStub(Class, NexusServiceOptions)} which has the
+     * highest precedence.
+     *
+     * @param nexusServiceOptions map from service to NexusServiceOptions
+     */
+    public Builder setNexusServiceOptions(Map<String, NexusServiceOptions> nexusServiceOptions) {
```
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/worker/WorkflowImplementationOptions.java:139]: No need to update doc, but just for me to understand, map key is service name here not class name right (even though class name is default service name)?

**Review by cretz — APPROVED** [review]
This is great. I looked over quite a bit looking for any obvious flaws and couldn't find any. Tests look good (and saw #2215). Would like @bergundy review before merge though.

File: temporal-sdk/src/main/java/io/temporal/worker/Worker.java (hunk 2 above)
Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/worker/Worker.java:424]: Correct

**Review by bergundy — APPROVED** [review]
Looks great!

The only major comment I had was about payload to content conversion but we discussed that on Zoom and the approach you took is okay for now. Functionally it does what we want, we will likely want to prevent the serialization though and build better support for this in the Nexus Java SDK.

File: temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+      } else {
+        startResponseBuilder.setAsyncSuccess(
+            StartOperationResponse.Async.newBuilder()
+                .setOperationId(result.getAsyncOperationId())
+                .build());
+      }
+    } catch (OperationUnsuccessfulException e) {
+      startResponseBuilder.setOperationError(
+          UnsuccessfulOperationError.newBuilder()
+              .setOperationState(e.getState().toString().toLowerCase())
```
[the remark this answers is not in the document]
Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java:214]: I made them lower case to match my understanding of the spec https://github.com/nexus-rpc/api/blob/main/SPEC.md#operationinfo

File: temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+              .setFailure(
+                  Failure.newBuilder()
+                      .setMessage("internal error")
+                      .setDetails(ByteString.copyFromUtf8(e.toString()))
+                      .build())
+              .build());
+    } finally {
+      // If the task timed out, we should not send a response back to the server
+      if (timedOut.get()) {
+        throw new TimeoutException("Nexus task complete after timeout.");
```
[the remark this answers is not in the document]
Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java:147]: >Also, is it standard to add punctuations to error messages in our SDK?

Yes most do

File: temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+          instance.getDefinition().getName(),
+          "\""
+              + instance.getDefinition().getName()
+              + "\" service type is already registered with the worker");
+    }
+  }
+
+  public CompletionStage<Void> shutdown(ShutdownManager shutdownManager, boolean unused) {
+    return shutdownManager.shutdownExecutorNow(
+        scheduler, "NexusTaskHandlerImpl#scheduler", Duration.ofSeconds(5));
```
[the remark this answers is not in the document]
Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java:247]: This is the executor to call the cancel the nexus operation context after its' timeout. By this point there shouldn't be any tasks to shutdown.

File: temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+              .setErrorType(e.getErrorType().toString())
+              .setFailure(createFailure(e.getFailureInfo()))
+              .build());
+    } catch (Throwable e) {
+      return new Result(
+          HandlerError.newBuilder()
+              .setErrorType(OperationHandlerException.ErrorType.INTERNAL.toString())
+              .setFailure(
+                  Failure.newBuilder()
+                      .setMessage("internal error")
```
[the remark this answers is not in the document]
Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java:140]: I do include the exceptions in this error, is there more we want to include?

**Comment by Quinn-With-Two-Ns** [comment]
Going to merge to unblock other PRs, will address any other comments in this PR in follow up PRs

File: temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (bergundy) [reply temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java:140]: I see it's in the details, can we copy the message instead?

File: temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bergundy) [reply temporal-sdk/src/main/java/io/temporal/internal/nexus/NexusTaskHandlerImpl.java:214]: I was suggesting that the enums in Java are made lower case so you don't have to do this transformation every time. Then the Nexus Java SDK would match the spec.

**Pull request merged**